core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java

```diff
@@ -171,7 +173,10 @@ public Set<Long> getReplicationLimitedFileIds() {
    * @return the root of the inode tree
    */
   public InodeDirectory getRoot() {
-    return mInodeStore.get(0).map(Inode::asDirectory).orElse(null);
+    if (mRoot == null) {
```

Comment:
The only way to properly avoid syncronized or volatile is to make sure you never see `mRoot==null`, rely on the writer to have updated it for you, and be a good reader to always read the cache. Alluxio does not allow updating root at runtime so once the master is started, this cache will never be updated. 

I'd like to start with assuming the cache is read here and do not try to update the cache. If you throw an error when root, is null, which will break?
1. Start alluxio from no journal
2. Start alluxio from existing journal
3. Run all UTs

Reply:
fter go through the old code, I agree it is impossible to return null while invoke getRoot for now,
since the root should be a directory all the time.